proto/v1beta1/intoto.proto

```diff
@@ -0,0 +1,145 @@
+message Signature {
+  string keyid = 1 [json_name="keyid"];
+  string sig = 2 [json_name="sig"];
+}
+
+// An object defined to capture the by products of a step.
+message ByProductObject {
+  int32 return_value = 1 [json_name="return-value"];
+  string stderr = 2 [json_name="stderr"];
+  string stdout = 3 [json_name="stdout"];
+}
+
+// This corrresponds to the core in-toto link attestation.
+message Link {
+  // This field contains the full command executed for the step. This can also
+  // be empty if attestations are generated for operations that aren't directly
+  // mapped to a specific command. Each term in the command is an independent
+  // string in the list. An example of a command in the in-toto metadata field
+  // is:
+  // "command": ["git", "clone", "https://github.com/in-toto/demo-project.git"]
+  repeated string command = 1 [json_name="command"];
+
+  // Materials are the supply chain artifacts that go into the step and are used
+  // for the operation performed. The key of the map is the path of the artifact
+  // and the structure contains the recorded hash information. An example is:
+  // "materials": {
+  //   "foo/bar": {
+  //     "sha256": "ebebf..."
+  //   }
+  // }
+  map<string, google.protobuf.Struct> materials = 2 [json_name="materials"];
+
+  // Products are the supply chain artifacts generated as a result of the step.
+  // The structure is identical to that of materials.
+  map<string, google.protobuf.Struct> products = 3 [json_name="products"];
+
+  // By Products are made up of the data that is generated as part of a software
+  // supply chain step, but are not the actual result of the step.
+  ByProductObject byproducts = 4 [json_name="byproducts"];
+
+  // This is a field that can be used to capture information about the
+  // environment. It is recommended for this field to contain information that
+  // details environment variables, filesystem information, and the present
+  // working directory. The structure of this field is:
+  // "environment": {
+  //   "variables": "<ENV>",
+  //   "filesystem": "<FS>",
+  //   "workdir": "<CWD>"
+  // }
+  map<string, google.protobuf.Struct> environment = 5 [json_name="environment"];
+}
+
+// This corresponds to a full signed attestation - it is made up of one or more
+// signatures and the core link attestation itself. This is used for occurrences
+// of a Grafeas in-toto note.
+message Details {
+  repeated Signature signatures = 1 [json_name="signatures"];
+  Link link = 2 [json_name="signed"];
+}
+
+// This contains the private and public portions of a key. The private portion
+// must be empty for a public key.
+message KeyValue {
+  string private = 1;
+  string public = 2;
+}
+
+// This defines the format used to record keys used in the software supply
+// chain. An in-toto link is attested by a functionary using one or more keys
+// defined in the in-toto layout.
+message FunctionaryKey {
```

Comment:
It seems like there should be some overlap with the `Signature` proto in `common.proto` here. Is it missing something you need to represent in `FunctionaryKey`?

Reply:
Note: I'm renaming `FunctionaryKey` to `SigningKey` to avoid confusions with the term Functionary.

As for overlap, as I understand, `Signature` defines a specific signature made using a particular key. `SigningKey` is a definition of such a key which can then be used to verify signatures. So I think they need to be distinct.